Support NODE_PATH in forked process

ingro opened the pull request:
This PR tries to fix #515 in a non-destructive way.

Basically it checks for the presence of `NODE_PATH` env variable before forking the process and creates an absolute path from it, so it could be used in `test-worker.js` as additional node module path.

That won't work there because `process.cwd()` points to the forked process test file.

novemberborn commented:
Rather than modifying `process.env.NODE_PATH` in `api.js` you should probably pass a custom value in `fork.js`: https://github.com/sindresorhus/ava/blob/bdb8b6ebaa05f1dc31cdab7c84aed7d92317b8ed/lib/fork.js#L19:L22

novemberborn commented on lib/test-worker.js:
```diff
@@ -86,6 +86,9 @@ var oldNodeModulesPaths = module.constructor._nodeModulePaths;
 	var ret = oldNodeModulesPaths.apply(this, arguments);
 	ret.push(nodeModulesDir);
+	if (process.env.NODE_PATH) {
+		ret.push(process.env.NODE_PATH);
```
`NODE_PATH` can contain multiple values. Not sure if the env var needs to be split and concatenated with `ret` here. See <https://nodejs.org/api/modules.html#modules_loading_from_the_global_folders> for more.

ingro commented:
Thanks for your input @novemberborn , I made the following changes:
- removed the ugly process.env.NODE_PATH modify in `api.js`
- added an option for the forked process (`projectRoot`) which contains the original path from where ava was called;
- cycled through possible multiple values of process.env.NODE_PATH (with a simple OS check) and join their path with the one supplied by `projectRoot`.

novemberborn commented on lib/fork.js:
```diff
@@ -13,7 +13,8 @@ module.exports = function (file, opts) {
 			rows: process.stdout.rows
-		} : false
+		} : false,
+		projectRoot: process.cwd()
```
This isn't necessarily the project root. Perhaps `parentWorkingDirectory`? Bit verbose though.

ingro replied:
What about `parentCwd`?

novemberborn replied:
Sure

novemberborn commented on lib/test-worker.js:
```diff
@@ -86,6 +86,12 @@ var oldNodeModulesPaths = module.constructor._nodeModulePaths;
 	var ret = oldNodeModulesPaths.apply(this, arguments);
 	ret.push(nodeModulesDir);
+	if (process.env.NODE_PATH) {
+		var osSplitChar = /^win/.test(process.platform) ? ';' : ',';
```
Perhaps use [`is-windows`](https://www.npmjs.com/package/is-windows). A separate package is likely preferable over inline regex fu.

---

> If the NODE_PATH environment variable is set to a colon-delimited list of absolute paths, then Node.js will search those paths for modules if they are not found elsewhere. (Note: On Windows, NODE_PATH is delimited by semicolons instead of colons.)

[(source)](https://nodejs.org/api/modules.html#modules_loading_from_the_global_folders)

The non-Windows separators are colons, not commas.

ingro replied:
I didn't want to add another dependency, but if it's fine for you I think too that this would be a better solution.

novemberborn replied:
I think it is, but I don't have final say ;-)

vdemedes replied:
No need to use `is-windows`, since all of its logic is 1 line (https://github.com/jonschlinkert/is-windows/blob/master/index.js#L20):

```js
process.platform === 'win32'
```

novemberborn replied:
Heh yea, fair enough.

ingro replied:
So I guess I could leave the test as it is (except for the wrong separators).

novemberborn replied:
Use `process.platform === 'win32'` though rather than a regular expression.

novemberborn commented on lib/test-worker.js:
```diff
@@ -86,6 +86,12 @@ var oldNodeModulesPaths = module.constructor._nodeModulePaths;
+	if (process.env.NODE_PATH) {
+		var osSplitChar = /^win/.test(process.platform) ? ';' : ',';
+		process.env.NODE_PATH.split(osSplitChar).forEach(function (additionalPath) {
+			ret.push(path.join(opts.projectRoot, additionalPath));
```
Should use [`path.resolve`](https://nodejs.org/api/path.html#path_path_resolve_from_to) in case `additionalPath` is already absolute.

ingro replied:
Yeah you're right, I'm really bad at this.

novemberborn replied:
No worries! This stuff is maddeningly difficult.

novemberborn commented:
Perhaps resolve the additional, absolute `NODE_PATH`s in the main process, once, and then pass the array through the `opts` to the `test-worker`.

You should add a test, probably in `test/fork.js`. Note that tests are written using `node-tap`. Maybe use [`t.spawn`](http://www.node-tap.org/advanced/#t-spawn-command-arguments-options-name-extra) to spawn another test file (from the `fixtures` directory) with a `NODE_PATH` environment variable, then in that file use the `lib/fork` module to run an AVA test. You could set up `NODE_PATH` so the final test can do an `import` and have it be resolved correctly. I'm in Gitter if that wasn't clear https://gitter.im/sindresorhus/ava

novemberborn commented on lib/fork.js:
```diff
@@ -13,7 +13,8 @@ module.exports = function (file, opts) {
 			rows: process.stdout.rows
-		} : false
+		} : false,
+		additionalPaths: opts.additionalPaths
```
`Object.assign()` already takes care of mixing in the `opts`.

novemberborn commented on lib/test-worker.js:
```diff
@@ -86,6 +86,11 @@ var oldNodeModulesPaths = module.constructor._nodeModulePaths;
+		opts.additionalPaths.forEach(function (additionalPath) {
+			ret.push(additionalPath);
+		});
+	}
```
Could replace with:

```js
return ret.concat(nodeModulesDir, additionalPaths);
```

novemberborn commented on cli.js:
```diff
@@ -93,11 +93,20 @@ if (cli.flags.init) {
+if (process.env.NODE_PATH) {
+	var osSplitChar = process.platform === 'win32' ? ';' : ':';
+	process.env.NODE_PATH.split(osSplitChar).forEach(function (additionalPath) {
+		additionalPaths.push(path.resolve(opts.projectRoot, additionalPath));
```
`process.cwd()` rather than `opts.projectRoot` right?

Perhaps `nodePaths` as the variable / option name? And personally I'd do something like:

```js
var nodePaths
if (process.env.NODE_PATH) {
	var osSplitChar = process.platform === 'win32' ? ';' : ':';
	nodePaths = process.env.NODE_PATH.split(osSplitChar).map(function (p) {
		return path.resolve(process.cwd(), p)
	});
} else {
	nodePaths = []
}
```

ingro replied:
Yep map is surely more elegant!

novemberborn commented on test/cli.js:
```diff
@@ -15,7 +15,10 @@ function execCli(args, dirname, cb) {
 
-	var env = {};
+	var env = {
+		// This probably should be set only for the corresponding test
```
Yep :)

ingro replied:
Any suggestion on how to approach this?

novemberborn replied:
Probably by adding an `env` argument after `dirname` and before `cb`. It should be optional though, no changes to the other tests should be necessary.

Could leave `dirname` required if `env` is provided, so `execCli([], { NODE_PATH: 'foo' })` is invalid.

novemberborn commented on test/cli.js:
```diff
@@ -110,3 +113,10 @@ test('pkg-conf: cli takes precedence', function (t) {
+		t.notOk(err);
+		t.end();
+	});
+});
```
Could you insert a linebreak at the end of the file?

novemberborn commented on test/fixture/node-paths/modules/nested/foo.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = function() {
+	return 'bar';
+}
```
Needs a semi-colon after the `}`. And could you insert a linebreak at the end of the file?

novemberborn commented on test/cli.js:
```diff
@@ -15,7 +15,10 @@ function execCli(args, dirname, cb) {
-	var env = {};
+	var env = {
+		// This probably should be set only for the corresponding test
+		NODE_PATH: 'node-paths/modules'
```
Should test with multiple paths, using the correct separator based on platform.

ingro commented:
I added the test and it seems like it's green now, but I surely forgot something else, let me know if something is missing!

shinnn commented on cli.js:
```diff
@@ -93,11 +94,22 @@ if (cli.flags.init) {
 
+var nodePaths;
+if (process.env.NODE_PATH) {
+    var osSplitChar = process.platform === 'win32' ? ';' : ':';
```
Instead you can use `path.delimiter`. https://nodejs.org/api/path.html#path_path_delimiter

ingro replied:
Cool! So I can get rid of all the env check stuff!

shinnn replied:
I should have commented it earlier :)

shinnn commented on cli.js:
```diff
@@ -93,11 +94,22 @@ if (cli.flags.init) {
+        return path.resolve(process.cwd(), p)
+    });
+} else {
+    nodePaths = []
```
```diff
-    nodePaths = []
+    nodePaths = [];
```

novemberborn commented:
LGTM!

ingro commented:
Thanks @novemberborn for your patience, I hope that if I'll ever have to contribuite again it will be a smoother process, I learned a lot anyway :)

novemberborn commented:
> Thanks @novemberborn for your patience, I hope that if I'll ever have to contribuite again it will be a smoother process, I learned a lot anyway :)

No worries, looking forward to your next contribution!

novemberborn commented on cli.js:
```diff
@@ -93,11 +94,21 @@ if (cli.flags.init) {
+var nodePaths;
+if (process.env.NODE_PATH) {
+    nodePaths = process.env.NODE_PATH.split(path.delimiter).map(function (p) {
+        return path.resolve(process.cwd(), p);
```
Just realized that `path.resolve(p)` defaults to `path.resolve(process.cwd(), p)` if `p` is relative. In other words you don't need to provide `process.cwd()` here :boom:

ingro replied:
There's always to learn something!

sindresorhus commented on lib/test-worker.js:
```diff
@@ -86,6 +86,9 @@ var oldNodeModulesPaths = module.constructor._nodeModulePaths;
 module.constructor._nodeModulePaths = function () {
 	var ret = oldNodeModulesPaths.apply(this, arguments);
 	ret.push(nodeModulesDir);
+	if (opts.nodePaths && opts.nodePaths.length > 0) {
```
`opts.nodePaths.length > 0` is not needed as it's handled by `concat`

sindresorhus commented on test/cli.js:
```diff
@@ -7,7 +7,7 @@ var getStream = require('get-stream');
 var cliPath = path.join(__dirname, '../cli.js');
 
-function execCli(args, dirname, cb) {
+function execCli(args, dirname, env, cb) {
```
Just replace `dirname` and `env` with an options object.

sindresorhus commented on test/cli.js:
```diff
@@ -110,3 +115,12 @@ test('pkg-conf: cli takes precedence', function (t) {
 });
+
+test('handles NODE_PATH', function (t) {
+	var nodePaths = 'node-paths/modules' + path.delimiter + 'node-paths/deep/nested';
```
Use `path.join()` instead

novemberborn replied:
@sindresorhus do you mean `path.join('node-paths', 'modules')`? `path.delimiter` is used for building the environment variable itself, which comprises multiple paths.

sindresorhus replied:
Oh, nvm, I read it as `path.sep`...

novemberborn replied:
Node probably handles any platform issues with the slashes in `node-paths/modules`.

sindresorhus commented on test/fixture/node-paths/deep/nested/path/bar.js:
```diff
@@ -0,0 +1,4 @@
+module.exports = function() {
```
`module.exports = function () {`

sindresorhus commented on test/fixture/node-paths/modules/nested/foo.js:
```diff
@@ -0,0 +1,4 @@
+module.exports = function() {
```
`module.exports = function () {`

sindresorhus commented on cli.js:
```diff
@@ -93,11 +94,21 @@ if (cli.flags.init) {
+    });
+} else {
+    nodePaths = [];
+}
```
```js
var nodePaths = process.env.NODE_PATH ? process.env.NODE_PATH
	.split(path.delimiter)
	.map(function (p) {
        return path.resolve(p);
    }) : [];
```

sindresorhus commented:
Ugh, I really dread this change. Not the PR itself, but the fact of having to support `NODE_PATH`, which is deprecated and awful in so many ways...

Wouldn't it be better to just pass all environment variables from the parent to the worker? That way we wouldn't have to special-case future environment variable needs.

novemberborn commented:
> Wouldn't it be better to just pass all environment variables from the parent to the worker? That way we wouldn't have to special-case future environment variable needs.

The problem is that `NODE_PATH` can be relative and the tests run in a different working directory.

sindresorhus commented:
```js
if (process.env.NODE_PATH) {
	process.env.NODE_PATH = process.env.NODE_PATH
		.split(path.delimiter)
		.map(function (p) {
			return path.resolve(p);
		})
		.join(path.delimiter);
}
```

Then?

novemberborn commented:
In `cli.js`? Anywhere else and it may interfere with the programatic API.

sindresorhus commented:
Yes, or even better, just pass this directly to the `env` option when forking the worker, without overwriting the current `process.env.NODE_PATH`.

ingro commented:
Well I will wait to make further changes so, while a better solution is found!

novemberborn commented:
@ingro I think what @sindresorhus is saying is that you can [add an `env` object in `fork.js`](https://github.com/sindresorhus/ava/blob/9c22ee4ac4384352f4c30ba7de56eac919f435ff/lib/fork.js#L21) which sets `NODE_PATH` to the corrected value. Then you don't need the changes in `test-worker.js`.

sindresorhus commented:
Yup, and keep in mind that the whole existing `process.env` needs to be passed to the fork thing to preserve existing behavior, so use `object-assign` to copy the object and then modify the `NODE_PATH` key.

ingro commented:
Well I think that's cleaner, I will look at it as soon as I have time!

ingro commented:
The commit has some conflicts, what's the procedure in situation like this?

sindresorhus commented:
@ingro You [resolve the merge conflict](https://www.google.co.th/search?q=git+resolve+merge+conflict&oq=git+resolve+merge+&aqs=chrome.0.0j69i57j0l4.5662j0j7&sourceid=chrome&es_sm=91&ie=UTF-8). (No need to open a new PR)

ingro commented:
It seems we are back to green!

novemberborn commented on lib/fork.js:
```diff
@@ -16,9 +16,25 @@ module.exports = function (file, opts) {
+			.map(function (p) {
+				return path.resolve(p);
+			})
+			.join(path.delimiter);
```
Could compute this outside of the function so it can be reused for subsequent forks.

sindresorhus commented on lib/fork.js:
```diff
@@ -7,6 +7,21 @@ var debug = require('debug')('ava');
+
+	env = objectAssign({
+		NODE_PATH: nodePath
+	}, env);
```
Ugh, discovered this too late. This makes no sense. Here you overwrite the modified `NODE_PATH` with the original one. This means the tests are also invalid. Tests still pass when the `.map` is commented out.

// @ingro @novemberborn

ingro commented:
Welp sorry, I blindly copied from one of your comment above.
That should fix the wrong assignment:
`env = objectAssign(env, { NODE_PATH: nodePath });`

Anyway tests still fails, will look into that!

ingro commented:
I think I found the problem with the test, it was:

```javascript
var nodePaths = 'node-paths/modules' + path.delimiter + 'node-paths/deep/nested';
```

while it should have been:

```javascript
var nodePaths = 'fixture/node-paths/modules' + path.delimiter + 'fixture/node-paths/deep/nested';
```

That's because `NODE_PATH` should be relative to the directory from which `ava` is running, not to the test itself.

I've fixed that in my fork, I should create another pull request?

novemberborn commented:
> I should create another pull request?

Yep! :dancers:

ingro commented:
Created here https://github.com/sindresorhus/ava/pull/559

hildjj commented:
@sindresorhus I think this regressed with commit 1f10a7cfa1b2ccc9378143c08f79727e725c5550.

novemberborn commented:
@hildjj that was intentional.

hildjj commented:
Is this the correct place to discuss whether that was the correct decision?  I don't see an issue or pull request associated with that commit, so I'll assume yes, unless you'd like me to open a new issue.

I have an (admittedly crappy) use case where being able to manipulate NODE_PATH is very handy.  It involves using the same test suite to test two different builds of the same software that have subtle differences in their dependencies.  I can figure out another way around this if I have to, but it was convenient.

novemberborn commented:
You could use an `ava.config.cjs` file to rewrite the `NODE_PATH` and configure it as an environment variable for AVA to use for the worker processes. We're not going to bring back specialized handling in AVA itself.

hildjj commented:
I'm arguing that it's not specialized handling.  ava is now modifying the environment being passed to my tests in a surprising way.  At the very least I would have expected something in the release notes and something in the docs for a change that breaks existing functionality built in to node.  The [docs](https://github.com/avajs/ava/blob/main/docs/06-configuration.md) currently say:

"environmentVariables: specifies environment variables to be made available to the tests. The environment variables defined here override the ones from process.env"

which is at least confusing at the moment, since it's not clear that the `process.env` isn't the same as the environment passed in to ava.

`ava.config.cjs` might work for me, with some logic to detect which of the cases I'm in, I suppose.

novemberborn commented:
No, AVA 3 no longer modifies anything. But let's continue this in a [Discussion](https://github.com/avajs/ava/discussions).

hildjj commented:
https://github.com/avajs/ava/discussions/2682 for linkage
